src/goto-instrument/goto_instrument_parse_options.cpp

```diff
@@ -257,6 +258,19 @@ int goto_instrument_parse_optionst::doit()
       return CPROVER_EXIT_SUCCESS;
     }
 
+    if(cmdline.isset("insert-final-assert-false"))
+    {
+      log.status() << "Inserting final assert false" << messaget::eom;
+      bool success = do_insert_final_assert_false(
```

Comment:
Remove `do_` prefix from name (already has a verb)

Reply:
@smowton Thanks for the feedback. I fixed the name of the function as suggested. Regarding calling the switch `--assert-function-does-not-terminate`. So if I have this correct if the assert passes (resp. fails) then the function does not terminate (does terminate)? The difficulty here is the "does not terminate" reason could be for multiple reasons: unbounded loop, not enough loop unrollings, inconsistent assumptions.

Reply:
@smowton I'm not sure I agree with the choice of name, this is only about checking for inconsistent assumptions---not asserting that the function doesn't terminate.

If you'd like the switch to reflect what it's for, perhaps something like `--check-for-inconsistent-assumptions` or `--ensure-consistent-assumptions`? However these also seem somewhat confusing to me---if the assumptions are inconsistent, the false assertion will actually *pass* rather than fail, which is different to all the other `--check-*` flags.

Reply:
Hmm, even if that's what it's for, it'll still pass/fail according to whether the function contains an infinite loop or similar, right? It seems like the only concrete information you could ascertain is "assertion reachable (i.e., fails)" -> "function can somehow terminate, i.e. is not impeded by inconsistent assumptions, unwind limit, depth limit, etc etc etc". In other words is this actually a version of `--cover`, with the specific coverage goal "end of specified function"?

Reply:
(actually not quite, since `--cover` would usually convert existing assertions to assumptions which this does not)

Reply:
Yes. The intention is that we would run CBMC on a binary instrumented with this flag, in parallel with an otherwise-identical binary that has not. Both CBMC runs would include exactly the same flags, including `--unwind` (which I believe one must always use on any nontrivial program anyway). So reachability and termination are for granted, unless the entire function that we're instrumenting is dead/unreachable.

This flag really is intended to show up contradictory assumptions, and nothing else. If you think there are other scenarios that might cause the assertion to be hit (or not hit), @nchong-at-aws should fix that. But if there are no such cases, then I think the flag should be precisely named to reflect either what it does (`--insert-final-assert-false`) or what it's for (`--ensure-consistent-assumptions`).

Reply:
So if the input program was
```
void f(int x) {
  assert(x == 5);
  assume(x % 2 == 0);
  assume(x % 2 == 1);
  // assert false introduced here
}
```
Then your non-instrumented binary would report failing the assertion `x == 5` and the instrumented one would fail that assertion but not the `assert(false)` introduced at the end? That would indicate that maybe there are contradictory assumptions (as there are in the example), but this example would behave the same, assuming a reasonable unwind limit:
```
void f(int x) {
  assert(x == 5);
  for(int i = 0; i < 1000000; ++i) { } // Presumably we have --unwind < 1000000
  // assert false introduced here
}
```
So I don't see how this plan disambiguates ability to reach the end due to assumptions vs. ability to reach the end due to unwind or depth limitation.

Reply:
I agree with you. There are multiple reasons why the instrumented assert could pass. This makes me lean towards the flag saying what it does and then leaving the "interpretation" of what to do with the result with the user. The way I imagine using this flag is after the harness has all-passes wrt its check-* and user-provided asserts.

Reply:
Fair enough, let's go with that.

Reply:
Thanks! I'll close this conversation.